File: 76-minimum-window-substring/Solution.java
import java.util.HashMap;

class Solution {

    public static void main(String[] args) {

        String s = "ADOBECODEBANC";
        String p = "ABC";
        System.out.println(minWindow(s, p));
    }

    public static String minWindow(String s, String p) {

        if (p.length() > s.length()) return "";

        HashMap<Character, Integer> map = new HashMap<>();
        for (char c : p.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        int begin = 0, end = 0, head = 0;
        int count = map.size();
        int minLen = Integer.MAX_VALUE;

        while (end < s.length()) {

            char c = s.charAt(end);

            if (map.containsKey(c)) {
                map.put(c, map.get(c) - 1);
                if (map.get(c) == 0) {
                    count--;
                }
            }
            end++;

            while (count == 0) {

                if (end - begin < minLen) {
                    head = begin;
                    minLen = end - begin;
                }

                c = s.charAt(begin);

                if (map.containsKey(c)) {
                    map.put(c, map.get(c) + 1);
                    if (map.get(c) > 0) {
                        count++;
                    }
                }
                begin++;
            }
        }

        return minLen == Integer.MAX_VALUE ? "" : s.substring(head, head + minLen);
    }
}

File: 76-minimum-window-substring/Solution.py
import sys


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        
        if(len(t) > len(s)): return ""

        n = len(s)

        map = {}

        for ch in t:
            if ch in map:
                map[ch] += 1
            else:
                map[ch] = 1
        
        print(map)

        end, begin, head = 0,0,0
        count = len(map)
        minLen = sys.maxsize

        while end < n:

            c = s[end]

            if(c in map):
                map[c] -= 1
                if map[c] == 0:
                    count -= 1

            end += 1
            print(map)
            print(begin, head, end)
            while count == 0:

                curLen = end - begin
                if (curLen < minLen):
                    head = begin
                    minLen = curLen
                
                c = s[begin]

                if c in map:
                    map[c] += 1
                    if map[c] > 0:
                        count += 1

                begin += 1

        if minLen == sys.maxsize:
            return ""
        else:
            return s[head:head + minLen] 



obj = Solution()

res = obj.minWindow("AA", "AA")

print(res)

